feat(iroh): Gossip client

rklaehn opened the pull request:
## Description

This makes gossip available in iroh. Using iroh-gossip directly, while not horrible, is a bit verbose.

## Breaking Changes

~~Not sure. It exports a few more things, and adds things. But in theory it should not modify existing things.~~

There should be none, and the semver test seems to agree...

## Notes & open questions

- ~~There can be some scenarios where this can cause trouble. E.g. when subscribing and then unsubscribing to a topic that is also used for doc sync.~~

This ^ is taken care of, since docs no longer use subscribe_all!

How to deal with missed messages due to slow reader:

I think that relying on guaranteed delivery for gossip is not a good idea in any case. so maybe just sending a Lagged but then continuing is best? Forcing the client to resubscribe could be a bit annoying.

## Change checklist

- [x] Self-review.
- [x] Documentation updates if relevant.
- [x] Tests if relevant.
- [x] All breaking changes documented.

rklaehn commented:
I *think* it is safe to just ignore any messages that are not for us. See https://github.com/n0-computer/iroh/pull/2258/commits/b88398e743935c5975571e8b4fcb54c929ddc5e8

Obviously you can mess with things by subscribing to a namespace that is currently being synced, but doing so accidentally is highly unlikely. And you can also mess with other stuff by using the low level blobs api, so 🤷 .

rklaehn commented:
It seems that due to the fact that the rpc protocol is public, we can not ever do an additional rpc request without breaking compat.

```
--- failure constructible_struct_adds_field: externally-constructible struct adds field ---

Description:
A pub struct constructible with a struct literal has a new pub field. Existing struct literals must be updated to include the new field.
        ref: https://doc.rust-lang.org/reference/expressions/struct-expr.html
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.31.0/src/lints/constructible_struct_adds_field.ron

Failed in:
  field Iroh.gossip in /home/runner/work/iroh/iroh/iroh/src/client.rs:44

--- failure enum_variant_added: enum variant added on exhaustive enum ---

Description:
A publicly-visible enum without #[non_exhaustive] has a new variant.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#enum-variant-new
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.31.0/src/lints/enum_variant_added.ron
```

dignifiedquire commented:
we need to mark the rpc enums as non exhaustive, that should solve this

dignifiedquire commented:
longterm I wish the rpc types where all private anyway

dignifiedquire commented:
> It seems that the sync engine thinks that it "owns" gossip, so it uses subscribe_all

@Frando can we change this? seems unfortunate to subscribe to everything by default

Frando commented:
> > It seems that the sync engine thinks that it "owns" gossip, so it uses subscribe_all
> 
> @Frando can we change this? seems unfortunate to subscribe to everything by default

I think we can, yes. What is our primitive of choice these days to merge a changing list of streams?

dignifiedquire commented:
- https://docs.rs/futures-lite/latest/futures_lite/stream/trait.StreamExt.html#method.race

Frando commented:
#2265 changes the sync engine usage of gossip to not use subscribe_all (based on this branch).
However as outlined in the previous comment I think I'd prefer to move the new gossip_dispatcher module to iroh_gossip, and make both the node's gossip client and the node's sync engine use that dispatcher.

rklaehn commented:
@Frando so the basic idea of the gossip dispatcher is just to have a thing that wraps the gossip and does some management so that it is less fragile to use.

If you agree with / can live with the public interface (which is just the subscribe function), we could move something with that signature into iroh-gossip as a higher level interface and then I would use it here, and you could update docs to use it (or not...).

The interface is basically this:

```rust
    pub fn subscribe(
        &self,
        msg: GossipSubscribeRequest,
        updates: UpdateStream,
    ) -> impl Stream<Item = RpcResult<GossipSubscribeResponse>> {
```

where UpdateStream is just a stream of either broadcast or broadcast_neighbours:

```rust
/// Send a gossip message
#[derive(Serialize, Deserialize, Debug)]
pub enum GossipSubscribeUpdate {
    /// Broadcast a message to all nodes in the swarm
    Broadcast(Bytes),
    /// Broadcast a message to all direct neighbors
    BroadcastNeighbors(Bytes),
}
```

Frando commented:
I think the API is fine. And still very much in favor of having only one and not 2 high level dispatch interfaces to gossip.

rklaehn commented:
Ok, I will clean up the dispatcher a bit and make sure it is independent of the rpc protocol, in preparation for moving it into iroh-gossip.

rklaehn commented:
Moved the dispatcher to iroh-gossip and moved the types into the dispatcher file.

I reexported some of the dispatcher types instead of making newtypes for them in the rpc protocol. Not 100% sure about this, but making so many single element newtypes is a drag...

rklaehn commented:
This is missing adapting the docs stuff to the new gossip dispatcher.

Frando commented:
Also at least a smoke test for the gossip client API would be good so that we don't break things.

rklaehn commented on iroh-gossip/src/dispatcher.rs:
```diff
@@ -0,0 +1,431 @@
+    /// This should not fail unless the gossip instance is faulty.
+    async fn dispatch_loop(self) -> anyhow::Result<()> {
+        use futures_lite::stream::StreamExt;
+        let stream = self.gossip.clone().subscribe_all();
```
Yeah, I did not look into the details of gossip that much so far. Just treated it as a black box. I guess once the dispatcher is moved to gossip we can optimize things. The important thing for me (and the reason for this PR in the first place) is to have a more high level way to use gossip.

rklaehn commented on iroh-gossip/src/dispatcher.rs:
```diff
@@ -0,0 +1,431 @@
+        /// Set of bootstrap nodes we are using.
+        bootstrap: BTreeSet<NodeId>,
+        /// The task that is driving the join future.
+        #[allow(dead_code)]
```
It's an AbortingJoinHandle, so it only exists for Drop. But somehow rust does not consider that during dead code analysis. Any way to make it aware?

rklaehn commented on iroh-net/src/portmapper/current_mapping.rs:
```diff
@@ -184,9 +184,6 @@ mod tests {
     // for testing a mapping is simply an ip, port pair
     type M = (Ipv4Addr, NonZeroU16);
 
-    const TEST_PORT: NonZeroU16 = // SAFETY: it's clearly non zero
```
Somehow I had dead code warnings without it 🤷

rklaehn commented on iroh-cli/src/commands/gossip.rs:
```diff
@@ -0,0 +1,71 @@
+                verbose,
+            } => {
+                let bootstrap = bootstrap.into_iter().collect();
+                let topic = blake3::hash(topic.as_ref()).into();
```
Not sure. Maybe we should just allow people to specify the exact gossip topic instead of hiding it. I was just too lazy to add a new command line arg.

Change the RPC to contain an actual topic-id, and have alternative CLI options topic: String (gets hashed on the client side) and --topic_raw: TopicId ?

rklaehn commented:
@Frando I have added a test. It seems to work, now that I have added the ability to tell the endpoint about node addrs (see https://github.com/n0-computer/iroh/pull/2433 ).

I think this kind of test is nice, because it forces us to make the client API sufficiently rich that you can do tests with it. Also it is much less likely to fluke out than a full cli type tests.

But - do you think this test for gossip will reliably work? I am connecting two nodes and then send a msg from one to the next. Can there be a situation where I send the message too early and it gets lost?

matheus23 commented on iroh/tests/client.rs:
```diff
@@ -0,0 +1,89 @@
+    Ok(())
+}
+
+// An uninhabited error type that allows us to use `?` in tests instead of `unwrap`.
+//
+// Any use of `?` in a test will immediately panic with the error message.
+#[derive(Debug)]
+enum TestError {}
+
+type TestResult<T = ()> = Result<T, TestError>;
+
+impl<E: std::fmt::Debug + std::fmt::Display + Send + Sync + 'static> From<E> for TestError {
+    fn from(error: E) -> Self {
+        panic!("Test failed: {:?}", error);
+    }
+}
```
How do we feel about using the crate for this? https://crates.io/crates/testresult

Been a happy user of it for a while. It's doing literally the same thing.
But we wouldn't have to redefine it for each test we're writing.

rklaehn replied:
Yeah, I didn't know about it. Just was annoyed by either having all the unwrap!() or having imprecise error locations.

Frando commented:
> But - do you think this test for gossip will reliably work? I am connecting two nodes and then send a msg from one to the next. Can there be a situation where I send the message too early and it gets lost?

Yes, Gossip::join is awaited here, and that waits for at least one neighbor to come up before completing.

So the remaining TODO is to merge the two dispatchers (in dispatcher.rs and net.rs) so that messages flow only over a single channel and not two. We should do this, but IMO it can also happen in a followup.
